### PR TITLE
reset refitting_level_ flag to false in error paths

### PR DESCRIPTION
**Pull request by ramvadiv**
Reset refitting_level_ flag to false in error paths in DBImpl::ReFitLevel()

### CODE REVIEW

**Comment by ajkr** [comment]
I restarted the failed jobs.

This test I've been trying to deflake for a while so will try fixing this failure:
```
[----------] Global test environment set-up.
[----------] 1 test from DBCompactionTestWithParam/DBCompactionTestWithParam
[ RUN      ] DBCompactionTestWithParam/DBCompactionTestWithParam.FlushAfterIntraL0CompactionCheckConsistencyFail/0
db/db_compaction_test.cc:5349: Failure
Expected equality of these values:
  i + 1
    Which is: 10
  NumTableFilesAtLevel(0)
    Which is: 5
[  FAILED  ] DBCompactionTestWithParam/DBCompactionTestWithParam.FlushAfterIntraL0CompactionCheckConsistencyFail/0, where GetParam() = (1, true) (202 ms)
[----------] 1 test from DBCompactionTestWithParam/DBCompactionTestWithParam (202 ms total)
```

This one's surprising. Maybe it is flaky:
```
[ RUN      ] DBOptionsTest.ChangeCompression
Failed to check Status 0x7ffc15331600
#0   ./db_options_test() [0x6bdd4f] rocksdb::port::PrintStack(int)	/home/travis/build/facebook/rocksdb/port/stack_trace.cc:121	
#1   ./db_options_test() [0x409d56] rocksdb::Status::~Status()	/home/travis/build/facebook/rocksdb/./include/rocksdb/status.h:43	
#2   ./db_options_test() [0x445613] rocksdb::Status::~Status()	??:?	
#3   ./db_options_test() [0x43c9d8] testing::AssertionResult::AssertionResult<bool>(bool const&, testing::internal::EnableIf<!testing::internal::ImplicitlyConvertible<bool, testing::AssertionResult>::value>::type*)	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest.h:19710 (discriminator 2)	
#4   ./db_options_test() [0x49db97] void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3959	
#5   ./db_options_test() [0x493f06] testing::internal::UnitTestImpl::os_stack_trace_getter()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6935	
#6   ./db_options_test() [0x494185] testing::Test::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3965	
#7   ./db_options_test() [0x494355] testing::TestInfo::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4124	
#8   ./db_options_test() [0x494835] testing::TestCase::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6685	
#9   ./db_options_test() [0x49e0d7] bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3959	
#10  ./db_options_test() [0x494b23] testing::UnitTest::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6245	
#11  ./db_options_test() [0x40d0ed] main	/home/travis/build/facebook/rocksdb/db/db_options_test.cc:938	
#12  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x1476d1eba830] ??	??:0	
#13  ./db_options_test() [0x427979] _start	??:?	
Received signal 6 (Aborted)
#0   /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x38) [0x1476d1ecf428] ??	??:0	
#1   /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a) [0x1476d1ed102a] ??	??:0	
#2   ./db_options_test() [0x409d5b] rocksdb::Status::~Status() [clone .part.41]	db_options_test.cc:?	
#3   ./db_options_test() [0x445613] rocksdb::Status::~Status()	??:?	
#4   ./db_options_test() [0x43c9d8] testing::AssertionResult::AssertionResult<bool>(bool const&, testing::internal::EnableIf<!testing::internal::ImplicitlyConvertible<bool, testing::AssertionResult>::value>::type*)	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest.h:19710 (discriminator 2)	
#5   ./db_options_test() [0x49db97] void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3959	
#6   ./db_options_test() [0x493f06] testing::internal::UnitTestImpl::os_stack_trace_getter()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6935	
#7   ./db_options_test() [0x494185] testing::Test::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3965	
#8   ./db_options_test() [0x494355] testing::TestInfo::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:4124	
#9   ./db_options_test() [0x494835] testing::TestCase::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6685	
#10  ./db_options_test() [0x49e0d7] bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:3959	
#11  ./db_options_test() [0x494b23] testing::UnitTest::Run()	/home/travis/build/facebook/rocksdb/third-party/gtest-1.8.1/fused-src/gtest/gtest-all.cc:6245	
#12  ./db_options_test() [0x40d0ed] main	/home/travis/build/facebook/rocksdb/db/db_options_test.cc:938	
#13  /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x1476d1eba830] ??	??:0	
#14  ./db_options_test() [0x427979] _start	??:?	
/bin/bash: line 1: 11020 Aborted                 (core dumped) ./$t
Makefile:1011: recipe for target 'check_some' failed
```

For testing, I actually did not find much tests for manual compaction failure scenarios (maybe that's why there's so many bugs). Here's one I wrote recently that might be a good starting point: https://github.com/ajkr/rocksdb/blob/46cf49c896dc4a53bee44c5ba807205610125e24/db/db_compaction_test.cc#L5445-L5537. I think the key difference is you want the L0->L1 to be issued _before_ the `ReFitLevel()` that does an L2->L1 in order to make the refitting compaction fail. Then, without this PR, any subsequent refitting compaction would also fail. After this PR, they should be able to succeed.

**Comment by ramvadiv** [comment]
This test runs fine in my devserver - but seems to fail pretty consistently on TravisCI. Will investigate more before pushing this thru...

**Comment by ajkr** [comment]
BTW, it's a bit hidden under all the test output, but just want to make sure it's clear the last paragraph in my above comment has a suggestion for how to test this bug fix. Admittedly, the suggestion requires the most effort per additional line covered I can imagine -- but we are really lacking in `CompactRange()` failure tests (I could only find that one :/ ) so in terms of feature coverage it'd be a big step forward.

**Comment by ajkr** [comment]
Sorry for spam. I also forgot again to mention we should add a release note in HISTORY.md under "Unreleased" "Bug fixes" section.

**Comment by ramvadiv** [comment]
Andrew - thanks a lot for the pointers/guidance on the test case. I have added a new test to exercise the RefitLevel() error path first and then retest the healthy path. Ran it with old RefitLevel() code and ensured it fails as expected and reran with the fix to ensure it passes.

File: HISTORY.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Rocksdb Change Log
 ## Unreleased
+### Bug fixes
+* Reset the `refitting_level_` flag to false in the error paths in `ReFitLevel()`. Without this fix, if ReFitLevel() returns error once, all subsequent calls to it will result in 'Not implemented: another thread is refitting' error.
```
Comment (ajkr) [inline HISTORY.md:4]: Proposed alternative in terms of the relevant user-level APIs.
```
* Fixed a bug after a `CompactRange()` with `CompactRangeOptions::change_level` set fails due to a conflict in the level change step, which caused all subsequent calls to `CompactRange()` with `CompactRangeOptions::change_level` set to incorrectly fail with a `Status::NotSupported("another thread is refitting")` error.
```

File: db/db_compaction_test.cc [hunk 2]
```diff
@@ -5719,6 +5718,81 @@ TEST_F(DBCompactionTest, ChangeLevelCompactRangeConflictsWithManual) {
...
+    // Select key range that matches the bottom most level (L2)
+    std::string begin_string = Key(0);
+    std::string end_string = Key(start_idx - 1);
+    Slice begin(begin_string);
+    Slice end(end_string);
+
+    CompactRangeOptions cro;
+    cro.change_level = true;
+    cro.target_level = 1;
+    ASSERT_NOK(dbfull()->CompactRange(cro, &begin, &end));
```
Comment (ajkr) [inline db/db_compaction_test.cc:5782]: Nice, I didn't notice the failure can be triggered without any sync points.

**Review by ajkr — APPROVED** [review]
Looks great! Just a suggestion on the release note.

File: HISTORY.md (hunk 1 above)
Comment (ramvadiv) [reply HISTORY.md]: Thanks Andrew. Fixed...

**Pull request closed without merging**